ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/TransactionCommitter.scala

```diff
@@ -123,31 +123,47 @@ private[kvutils] class TransactionCommitter(
   }
 
   /** Authorize the submission by looking up the party allocation and verifying
-    * that the submitting party is indeed hosted by the submitting participant.
+    * that all of the submitting parties are indeed hosted by the submitting participant.
     */
-  private def authorizeSubmitter: Step = (commitContext, transactionEntry) => {
-    commitContext.get(partyStateKey(transactionEntry.submitter)) match {
-      case Some(partyAllocation) =>
-        if (partyAllocation.getParty.getParticipantId == commitContext.participantId)
+  private[committer] def authorizeSubmitters: Step = (commitContext, transactionEntry) => {
+    def rejection(reason: RejectionReason) =
+      reject[DamlTransactionEntrySummary](
+        commitContext.recordTime,
+        buildRejectionLogEntry(transactionEntry, reason))
+
+    @scala.annotation.tailrec
+    def authorizeAll(submitters: List[Party]): StepResult[DamlTransactionEntrySummary] =
+      submitters match {
+        case Nil =>
           StepContinue(transactionEntry)
-        else
-          reject(
-            commitContext.recordTime,
-            buildRejectionLogEntry(
-              transactionEntry,
+        case submitter :: others =>
+          authorize(submitter) match {
+            case Some(rejection) =>
+              rejection
+            case None =>
+              authorizeAll(others)
+          }
+      }
```

Comment:
If you do this with a `.foldLeft`, you can do it on a `Set` (or any `Traversable`), which gets rid of a conversion.

Something like this should work.

```scala
      submitters.foldLeft(Right(StepContinue(transactionEntry))) {
        case Left(rejection) =>
          rejection
        case Right(result) =>
          authorize(submitter) match {
            case Some(rejection) => Left(rejection)
            case None => Right(result)
          }
      }
```

You can probably also do this with Scalaz's `traverse`.

Reply:
As far as I understand the `.foldLeft` approach doesn't allow to fail-fast - I've used the recursion specifically for that.
Should I change this?

Reply:
Leaving as it is then 👍